### PR TITLE
Feat/support server storage & ssr cache

### PR DESCRIPTION
**Pull request by GiveMe-A-Name**
## Summary

more detail reference to:
- https://github.com/web-infra-dev/modern.js/discussions/4968
- https://github.com/web-infra-dev/modern.js/discussions/4967

## Checklist

- [ ] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by GiveMe-A-Name** [comment]
!canary

File: packages/toolkit/runtime-utils/src/node/fileReader.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      // if is not exist, return null value.
+      this.storage.set(path, null);
+      return null;
+    }
+  }
+
+  /**
+   * Clear the fileCache entriely.
+   */
+  reset() {
```
Comment (GiveMe-A-Name) [inline packages/toolkit/runtime-utils/src/node/fileReader.ts:52]: 保证这里的 reset 是一个安全的移除。

File: packages/server/prod-server/src/libs/render/ssrCache/index.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  } else if (isCacheControl(cacheOption)) {
+    return cacheOption;
+  } else if (isCacheOptionProvider(cacheOption)) {
+    return cacheOption(req);
+  } else {
+    const url = req.url!;
+    const options = Object.entries(cacheOption);
+
+    for (const [key, option] of options) {
+      if (key === '*' || new RegExp(key).test(url)) {
```
Comment (GiveMe-A-Name) [inline packages/server/prod-server/src/libs/render/ssrCache/index.ts:62]: 考虑这里是否使用正则来进行匹配。

File: packages/toolkit/types/server/utils.d.ts [hunk 3]
```diff
@@ -69,3 +69,88 @@ export type BffProxyOptions =
   | Record<string, ProxyDetail>
   | ProxyDetail[]
   | ProxyDetail;
+
+export type CacheControlRange = 'header' | 'query';
+export type CacheControl = {
```
Comment (GiveMe-A-Name) [inline packages/toolkit/types/server/utils.d.ts:74]: 用更灵活和自由的方式

File: packages/toolkit/types/server/utils.d.ts [hunk 4]
```diff
@@ -69,3 +69,88 @@ export type BffProxyOptions =
...
+  req: IncomingMessage,
+) => Promise<CacheControl> | CacheControl;
+
+export type CacheOption =
+  | false
+  | CacheOptionProvider
+  | CacheControl
+  | Record<string, CacheControl | CacheOptionProvider>;
+
+export interface Container<K = string, V = string> {
```
[the remark this answers is not in the document]
Comment (GiveMe-A-Name) [reply packages/toolkit/types/server/utils.d.ts:110]: Done

File: packages/toolkit/types/server/utils.d.ts (hunk 3 above)
Comment (GiveMe-A-Name) [reply packages/toolkit/types/server/utils.d.ts:74]: Done

File: packages/document/main-doc/docs/zh/guides/advanced-features/ssr/cache.mdx [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+export type CacheOptions = Record<string, CacheControl | CacheOptionProvider>;
+```
+
+有时开发者面对不同的路由需要应用不同的缓存策略。我们也提供一种映射的方式进行配置, 以下列代码为例子
+
+```ts title="server/cache.ts"
+import type { CacheOption } from '@modern-js/runtime/server';
+
+export const cacheOption: CacheOption = {
+  '/home': {
```
[the remark this answers is not in the document]
Comment (GiveMe-A-Name) [reply packages/document/main-doc/docs/zh/guides/advanced-features/ssr/cache.mdx:90]: done~

File: packages/runtime/plugin-runtime/src/ssr/serverRender/renderToString/index.ts [hunk 6]
```diff
@@ -1,6 +1,5 @@
 import { run } from '@modern-js/runtime-utils/node';
 import { ServerRenderOptions } from '../types';
-import { PreRender } from '../../react/prerender';
```
[the remark this answers is not in the document]
Comment (GiveMe-A-Name) [reply packages/runtime/plugin-runtime/src/ssr/serverRender/renderToString/index.ts:3]: done~

File: tests/integration/ssr/fixtures/base/server/cache.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import type { CacheOption, Container } from '@modern-js/runtime/server';
```
[the remark this answers is not in the document]
Comment (GiveMe-A-Name) [reply tests/integration/ssr/fixtures/base/server/cache.ts:1]: done~, add a test unit

**Comment by GiveMe-A-Name** [comment]
!canary

**Comment by GiveMe-A-Name** [comment]
!canary

File: packages/document/main-doc/docs/en/guides/advanced-features/ssr/cache.mdx [hunk 8]
```diff
@@ -0,0 +1,179 @@
+---
+sidebar_position: 3
+title: Cache
+---
+
+# Render Cache
```
[the remark this answers is not in the document]
Comment (GiveMe-A-Name) [reply packages/document/main-doc/docs/en/guides/advanced-features/ssr/cache.mdx:6]: Done~

**Comment by GiveMe-A-Name** [comment]
!canary

**Review by yimingjfe — APPROVED** [review]

**Pull request merged**